src/openms/source/CONCEPT/QCChargeDistribution.cpp

```diff
@@ -0,0 +1,156 @@
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Maria Trofimova $
+// $Authors: Maria Trofimova $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/CONCEPT/QCChargeDistribution.h>
+#include <OpenMS/METADATA/PeptideIdentification.h>
+#include <OpenMS/KERNEL/FeatureMap.h>
+
+using namespace OpenMS;
+using namespace std;
+
+//Destructor
+QCChargeDistribution::~QCChargeDistribution()
+{
+
+}
+
+//Constructor
+QCChargeDistribution::QCChargeDistribution(std::vector<std::pair<OpenMS::String,OpenMS::FeatureMap>> files):
+  feat_map_(files)
+  {
+      
+  };
+  
+//Main method to write mztab peptide section data needed for charge distribution plot (PTXQC)
+int QCChargeDistribution::ChargeDistribution(MzTab& mztab) const
+{
+  vector<FeatureMap> maps;
+  MzTabPeptideSectionRows rows;
+  MzTabPeptideSectionRows mztabRows = mztab.getPeptideSectionRows();
+  int pepIDCount = 0;
+
+  for(vector<pair<String,FeatureMap>>::const_iterator it = feat_map_.begin();it!=feat_map_.end();++it)
+  {
+    maps.push_back (it->second);
+  }
+   
+  for (unsigned long m = 0; m < maps.size(); m++)
+  {     
+  
+    String rfile;
+    
+    if (maps[m].metaValueExists("spectra_data")) 
+    {		
+      StringList rfiles = maps[m].getMetaValue("spectra_data");	
+      rfile = rfiles[0];
+    } 
+    
+    for (vector<Feature>::const_iterator f_it = maps[m].begin(); f_it!=maps[m].end();f_it++)
+    {
+
+      vector<PeptideIdentification> pep_id = f_it->getPeptideIdentifications();	
+ 		  
+      if (pep_id.empty()) 
+      {
+        //Empty lines peptide_data_ for features with retention times	
+      }
+ 				
+      else 
+      {
+ 	    for (vector<PeptideIdentification>::iterator p_it = pep_id.begin(); p_it!=pep_id.end(); p_it++) 
+ 	    {
+ 	      pepIDCount++;
+ 	      MzTabPeptideSectionRow row;
+ 	      MzTabString PepSeq;
+ 	      MzTabInteger charge;
+ 		  
+ 		  //Set sequence and charge	
+          vector<PeptideHit> hits = p_it->getHits(); 
+          PeptideHit hit = hits[0];
+          AASequence seq = hit.getSequence();
+          int ch = hit.getCharge();
+          PepSeq.set(seq.toString());
+          charge.set(ch);
+          row.charge = charge;
+          row.sequence = PepSeq;
+          
+    	  //Set optional columns: original RT and source file
+          vector<MzTabOptionalColumnEntry> v;                
```

Comment:
this is duplicated code, right?!
You can leave the other columns empty and just add charge. Even if that means its only a single column.